WebViews: Show 'No Connection' alert if internet not reachable.

ScoutHarris opened the pull request:
Fixes #9284 

When a `WebView` is displayed and there is no internet connection, the 'No Connection' alert is now displayed:
<img width="366" alt="no_connection" src="https://user-images.githubusercontent.com/1816888/41560267-cd5aa4c8-7303-11e8-83d0-253bfc595513.png">

To test:
Disconnect internet. Verify the alert shows when accessing `WebView`s, for example from:
- [ ] Site > View Site.
- [ ] Me > App Settings > Privacy Settings > any link.
- [ ] Also, please test on iOS 11 and iOS 10 as there are different controllers depending on iOS version.

ScoutHarris commented:
Thanks for the info @koke ! See what you think about it now.

koke commented:
I ran into a couple issues while testing this, when I turn on Airplane Mode and go to view site, I get the alert, but when I dismiss it I just see this:

![screen shot 2018-06-20 at 12 18 49](https://user-images.githubusercontent.com/8739/41652686-1db0a9f0-7484-11e8-935f-def61a451bd5.png)

The reload/share/browser buttons are enabled but none of them do anything.

The second one is more puzzling. If I generate an error other than no connectivity (I took my web server down), the same thing happens, but then when I close the web view and tap on View Site again I get a blank view where nothing happens, and it looks like the call to `WKHTTPCookieStore.getAllCookies` is never calling its callback block.

koke commented on WordPress/Classes/Utility/WebKitViewController.swift:
```diff
@@ -354,4 +354,14 @@ extension WebKitViewController: WKUIDelegate {
+    func webView(_ webView: WKWebView, didFailProvisionalNavigation navigation: WKNavigation!, withError error: Error) {
+        DDLogInfo("\(NSStringFromClass(type(of: self))) Error Loading [\(error)]")
+
+        if !ReachabilityUtils.isInternetReachable() {
```
I noticed that when trying this on airplane mode, the message on `error` would be the exact same one provided by `ReachabilityUtils.showAlertNoInternetConnection()`, so we might not need to handle it as a special case

ScoutHarris replied:
That is true (except for the alert title). However, I did this so the no connection alert would be consistent. Since we have a specific `No Connection` alert, if it changes in the future this one will be in sync.

koke commented:
My guess at least for the first issue is that we're adding the alert on `didFailProvisionalNavigation` instead of `didFail`, but I haven't tested with that

ScoutHarris commented:
> My guess at least for the first issue is that we're adding the alert on `didFailProvisionalNavigation` instead of `didFail`


Well, `didFail` does not get called at this point, which is why I used `didFailProvisionalNavigation`. Per the [documentation](https://developer.apple.com/documentation/webkit/wknavigationdelegate?changes=latest_minor):
`didFail` - Called when an error occurs during navigation.
`didFailProvisionalNavigation` - Called when an error occurs while the web view is loading content.

I shall look at the issues you found. Thank you!

ScoutHarris commented:
Hey @koke . I believe I've addressed all the issues.

> The reload/share/browser buttons are enabled but none of them do anything.

The buttons are now disabled if there is no connection. It now observes network reachability and refreshes the webview when it changes.

Notes: 
  - In the simulator, sometimes the reachability notification took quite a while to trigger, and sometimes never would. On a device, it works as expected.
  - There is an issue where the webview buttons don't show in iOS 10 (https://github.com/wordpress-mobile/WordPress-iOS/issues/9572)

>  tap on View Site again I get a blank view where nothing happens

For some reason, `WPError` is stopping the webview from getting cookies. Sometimes. `ReachabilityUtils` alert did not cause this problem. So, instead of banging my head on that more, and possibly affecting other `WPError` usage, I added a new method to `ReachabilityUtils` to show a generic alert. 

Ready for another review please kind sir!

ScoutHarris commented:
Hey @jleandroperez . Since @koke is at a conference, would you mind reviewing this please? If you have any questions about the issues or what was done, feel free to ping me.

ScoutHarris commented:
Hey @koke , @jleandroperez .

I could use some assistance on this one. The issue both of you are seeing is that eventually getting cookies for the URL fails. Repeating these steps will work for several rounds, but will eventually fail, and the WebView will stop functioning:
1. Disable internet connection.
2. View Site.
3. OK on No Connection alert.
4. Close webview.
5. Repeat from step 2.


When I say getting cookies for the URL fails, I mean this method in `CookieJar.swift` returns nothing:

```
extension WKHTTPCookieStore: CookieJarSharedImplementation {
    func getCookies(url: URL, completion: @escaping ([HTTPCookie]) -> Void) {
        getAllCookies { (cookies) in
            DispatchQueue.main.async {
                completion(cookies.filter({ (cookie) in
                    return cookie.matches(url: url)
                }))
            }
        }
    }
...    
```





Showing the alert is eventually disrupting something. But I cannot figure out what or why. _Any_ assistance would be _greatly_ appreciated as I am stumped.

Thank you!!

cc @aerych

koke commented:
I can't test this until tomorrow, but some things I'd like to test in case you want to give it a try before I get to it:

- Disable the authentication for View Site and see if it still breaks. We're seeing the problem on `getAllCookies` but I think the whole WebKit service dies and no web view would work after what's breaking happens.
- See if we can reproduce on a minimal sample project, and file a radar about it if it seems like a WebKit/UIKit bug
- A smaller sample project also might help in coming up with a workaround if this is actually an iOS bug

koke commented:
I've been playing with this for a bit and just can't figure it out. I tried adding an async to the delegate method, since the alert was being presented from the WebKit runloop, but that didn't fix anything:

![screen shot 2018-07-04 at 11 57 52](https://user-images.githubusercontent.com/8739/42272689-40642c66-7f87-11e8-968c-1cd325cf4c6e.png)

It is indeed just the cookies method failing after whatever happens, if you try to load a webview without authentication (e.g. Privacy Policy) it will work.

My recommendation would be:

- Still try to see if we can isolate this issue in a small project and report to Apple.
- For a workaround, add some timeout mechanism to `getAllCookies` so it calls `completion([])` if nothing happens after a couple seconds.

ScoutHarris commented:
@koke @jleandroperez - ready for another review if you please.

> For a workaround, add some timeout mechanism to getCookies so it calls completion([]) if nothing happens after a couple seconds.

@koke thank you so much for your help and this suggestion!!! I did that, and it seems to work fine now.

ScoutHarris commented:
Thanks @jleandroperez !

Just to point out, I did note this previously (in a very early comment):

> There is an issue where the webview buttons don't show in iOS 10 (#9572)

koke commented:
The latest changes work great, thanks for all the hard work :shipit: 🎉 👏

ScoutHarris commented:
Thanks for all your help @koke !
